### PR TITLE
feat: Add the voice state event

### PR DESCRIPTION
**Pull request by sillyElvy**
## About

This pull request is about (X), which does (Y).

## Checklist

- [x] The ``pre-commit`` code linter has been run over all edited files to ensure the code is linted.
- [x] I've ensured the change(s) work on `3.8.6` and higher.

I've made this pull request: (check all that apply)
  - [ ] For the documentation
  - [x] To add a new feature
  - [ ] As a general enhancement
  - [ ] As a refactor of the library/the library's code
  - [ ] To fix an existing bug
  - [ ] To resolve #ISSUENUMBER

This is:
  - [ ] A breaking change

### CODE REVIEW

File: interactions/api/models/gw.py [hunk 1]
```diff
@@ -572,3 +573,103 @@ class Webhooks(DictSerializerMixin):
...
+
+    @property
+    def joined(self) -> bool:
+        """
+        Whether the user joined the channel.
+        :rtype: bool
+        """
+        return self.channel_id is not None
+
+    async def mute_member(self, reason: Optional[str]) -> Member:
```
Comment (Damego) [inline interactions/api/models/gw.py]:
```suggestion
    async def mute_member(self, reason: Optional[str] = None) -> Member:
```

File: interactions/api/models/gw.py [hunk 2]
```diff
@@ -572,3 +573,103 @@ class Webhooks(DictSerializerMixin):
...
+        """
+        Mutes the current member.
+        :param reason: The reason of the muting, optional
+        :type reason: str
+        :return: The modified member object
+        :rtype: Member
+        """
+        return await self.member.modify(guild_id=int(self.guild_id), mute=True, reason=reason)
+
+    async def deafen_member(self, reason: Optional[str]) -> Member:
```
Comment (Damego) [inline interactions/api/models/gw.py]:
```suggestion
    async def deafen_member(self, reason: Optional[str] = None) -> Member:
```

File: interactions/api/models/gw.py [hunk 3]
```diff
@@ -572,3 +573,103 @@ class Webhooks(DictSerializerMixin):
...
+        """
+        Deafens the current member.
+        :param reason: The reason of the deafening, optional
+        :type reason: str
+        :return: The modified member object
+        :rtype: Member
+        """
+        return await self.member.modify(guild_id=int(self.guild_id), deaf=True, reason=reason)
+
+    async def move_member(self, channel_id: int, *, reason: Optional[str]) -> Member:
```
Comment (Damego) [inline interactions/api/models/gw.py]:
```suggestion
    async def move_member(self, channel_id: int, *, reason: Optional[str] = None) -> Member:
```

File: interactions/api/models/gw.py [hunk 4]
```diff
@@ -572,3 +573,103 @@ class Webhooks(DictSerializerMixin):
...
+        :rtype: Channel
+        """
+        return Channel(**await self._client.get_channel(int(self.channel_id)), _client=self._client)
+
+    async def get_guild(self) -> "Guild":
+        """
+        Gets the guild in what the update took place.
+        :rtype: Guild
+        """
+        return Guild(**await self._client.get_guild(int(self.channel_id)), _client=self._client)
```
Comment (Damego) [inline interactions/api/models/gw.py]:
```suggestion
        return Guild(**await self._client.get_guild(int(self.guild_id)), _client=self._client)
```

**Review by Damego — CHANGES_REQUESTED** [review]

**Comment by Damego** [comment]
I would do like to see some useful properties from the lavalink ext 
image

File: interactions/api/models/gw.py [hunk 5]
```diff
@@ -572,3 +573,103 @@ class Webhooks(DictSerializerMixin):
...
+        Whether the user joined the channel.
+        :rtype: bool
+        """
+        return self.channel_id is not None
+
+    async def mute_member(self, reason: Optional[str]) -> Member:
+        """
+        Mutes the current member.
+        :param reason: The reason of the muting, optional
+        :type reason: str
```
Comment (Damego) [inline interactions/api/models/gw.py]: I think you could apply new docs style

File: interactions/api/models/gw.py [hunk 6]
```diff
@@ -572,3 +573,103 @@ class Webhooks(DictSerializerMixin):
...
+        converter=datetime.fromisoformat, default=None
+    )
+    suppress: Optional[bool] = field(default=None)
+    session_id: Optional[int] = field(default=None)
+    self_video: Optional[bool] = field(default=None)
+    self_mute: Optional[bool] = field(default=None)
+    self_deaf: Optional[bool] = field(default=None)
+    self_stream: Optional[bool] = field(default=None)
+    mute: Optional[bool] = field(default=None)
+    deaf: Optional[bool] = field(default=None)
```
Comment (Damego) [inline interactions/api/models/gw.py]: Why all attributes are optional?

Comment (sillyElvy) [reply interactions/api/models/gw.py]: Idk xD

File: interactions/api/models/member.py [hunk 7]
```diff
@@ -75,6 +76,18 @@ def __getattr__(self, name):
         except AttributeError as e:
             raise AttributeError(f"Neither `User` nor `Member` have attribute {name}") from e
 
+    @property
+    def voice_state(self) -> "VoiceState":
```
Comment (Damego) [inline interactions/api/models/member.py]:
```suggestion
    def voice_state(self) -> Optional["VoiceState"]:
```

File: interactions/api/models/member.py [hunk 8]
```diff
@@ -75,6 +76,18 @@ def __getattr__(self, name):
...
+        """
+        Returns the current voice state of the member, if any.
+        :rtype: VoiceState
+        """
+        if not self._client:
+            raise LibraryException(code=13)
```
Comment (Damego) [inline interactions/api/models/member.py:87]:
```suggestion
            raise LibraryException(code=13)

```

File: interactions/api/gateway/client.py [hunk 9]
```diff
@@ -601,6 +606,8 @@ def __get_object_id(
         """
         if isinstance(obj, (Member, GuildMember)):
             id = (Snowflake(data["guild_id"]), obj.id)
+        if isinstance(obj, VoiceState):
+            id = Snowflake(data["user_id"])
```
Comment (Damego) [inline interactions/api/gateway/client.py]:
```suggestion
            id = obj.user_id
```

File: interactions/api/gateway/client.py [hunk 10]
```diff
@@ -542,6 +543,10 @@ def _dispatch_event(self, event: str, data: dict) -> None:
                         old_obj = obj
 
                     _cache.add(old_obj, id)
+
+                    if event == "VOICE_STATE_UPDATE" and not data.get("channel_id"):  # user left
+                        del _cache[Snowflake(data.get("user_id"))]
```
Comment (Damego) [inline interactions/api/gateway/client.py]:
```suggestion
                        del _cache[obj.user_id]
```

File: interactions/api/gateway/client.py [hunk 11]
```diff
@@ -542,6 +543,10 @@ def _dispatch_event(self, event: str, data: dict) -> None:
                         old_obj = obj
 
                     _cache.add(old_obj, id)
+
+                    if event == "VOICE_STATE_UPDATE" and not data.get("channel_id"):  # user left
```
Comment (Damego) [inline interactions/api/gateway/client.py]:
```suggestion
                    if event == "VOICE_STATE_UPDATE" and not obj.channel_id:  # user left
```

File: interactions/api/models/guild.py [hunk 12]
```diff
@@ -540,6 +540,37 @@ async def ban(
...
 
+    @property
+    def voice_states(self) -> List["VoiceState"]:
+        """
+        Gets all voice states of the guild.
+        :rtype: List[VoiceState]
+        """
+
+        if not self._client:
+            raise LibraryException(code=13)
```
Comment (Damego) [inline interactions/api/models/guild.py:552]:
```suggestion
            raise LibraryException(code=13)

```

**Review by Damego — COMMENTED** [review]
bit more xD

File: interactions/api/models/gw.py [hunk 13]
```diff
@@ -572,3 +573,101 @@ class Webhooks(DictSerializerMixin):
...
+    deaf: bool = field()
+
+    def __attrs_post_init__(self):
+        if self.member:
+            self.member._extras["guild_id"] = self.guild_id
+
+    @property
+    def joined(self) -> bool:
+        """
+        Whether the user joined the channel.
```
Comment (Damego) [inline interactions/api/models/gw.py:624]:
```suggestion
        Whether the user joined the channel.

```

File: interactions/api/models/gw.py [hunk 14]
```diff
@@ -572,3 +573,101 @@ class Webhooks(DictSerializerMixin):
...
+        :rtype: bool
+        """
+        return self.channel_id is not None
+
+    async def mute_member(self, reason: Optional[str] = None) -> Member:
+        """
+        Mutes the current member.
```
Comment (Damego) [inline interactions/api/models/gw.py:632]:
```suggestion
        Mutes the current member.

```

File: interactions/api/models/gw.py [hunk 15]
```diff
@@ -572,3 +573,101 @@ class Webhooks(DictSerializerMixin):
...
+        :param Optional[str] reason: The reason of the muting, optional
+        :return: The modified member object
+        :rtype: Member
+        """
+        return await self.member.modify(guild_id=int(self.guild_id), mute=True, reason=reason)
+
+    async def deafen_member(self, reason: Optional[str] = None) -> Member:
+        """
+        Deafens the current member.
```
Comment (Damego) [inline interactions/api/models/gw.py:642]:
```suggestion
        Deafens the current member.

```

File: interactions/api/models/gw.py [hunk 16]
```diff
@@ -572,3 +573,101 @@ class Webhooks(DictSerializerMixin):
...
+        :return: The modified member object
+        :rtype: Member
+        """
+        return await self.member.modify(guild_id=int(self.guild_id), deaf=True, reason=reason)
+
+    async def move_member(
+        self, channel_id: Union[int, str, Snowflake], *, reason: Optional[str] = None
+    ) -> Member:
+        """
+        Moves the member to another channel.
```
Comment (Damego) [inline interactions/api/models/gw.py:654]:
```suggestion
        Moves the member to another channel.

```

File: interactions/api/models/gw.py [hunk 17]
```diff
@@ -572,3 +573,101 @@ class Webhooks(DictSerializerMixin):
...
+        :return: The modified member object
+        :rtype: Member
+        """
+        return await self.member.modify(
+            guild_id=int(self.guild_id), channel_id=channel_id, reason=reason
+        )
+
+    async def get_channel(self) -> Channel:
+        """
+        Gets the channel in what the update took place.
```
Comment (Damego) [inline interactions/api/models/gw.py:667]:
```suggestion
        Gets the channel in what the update took place.

```

File: interactions/api/models/gw.py [hunk 18]
```diff
@@ -572,3 +573,101 @@ class Webhooks(DictSerializerMixin):
...
+        :rtype: Channel
+        """
+        return Channel(**await self._client.get_channel(int(self.channel_id)), _client=self._client)
+
+    async def get_guild(self) -> "Guild":
+        """
+        Gets the guild in what the update took place.
```
Comment (Damego) [inline interactions/api/models/gw.py:675]:
```suggestion
        Gets the guild in what the update took place.

```

File: interactions/api/models/member.py [hunk 19]
```diff
@@ -75,6 +76,19 @@ def __getattr__(self, name):
         except AttributeError as e:
             raise AttributeError(f"Neither `User` nor `Member` have attribute {name}") from e
 
+    @property
+    def voice_state(self) -> Optional["VoiceState"]:
+        """
+        Returns the current voice state of the member, if any.
```
Comment (Damego) [inline interactions/api/models/member.py:82]:
```suggestion
        Returns the current voice state of the member, if any.

```

**Review by Damego — COMMENTED** [review]
Shouldn't be blank line between description and params?

File: interactions/api/models/channel.py [hunk 20]
```diff
@@ -500,6 +501,28 @@ def mention(self) -> str:
         """
         return f"<#{self.id}>"
 
+    @property
+    def voice_states(self) -> List["VoiceState"]:
+        """
+        Returns all voice states this channel has. Only applicable for voice channels.
```
Comment (Damego) [inline interactions/api/models/channel.py:507]:
```suggestion
        Returns all voice states this channel has. Only applicable for voice channels.

```

File: interactions/api/models/guild.py [hunk 21]
```diff
@@ -540,6 +540,38 @@ async def ban(
             if int(member.id) == _member_id:
                 return self.members.remove(member)
 
+    @property
+    def voice_states(self) -> List["VoiceState"]:
+        """
+        Gets all voice states of the guild.
```
Comment (Damego) [inline interactions/api/models/guild.py:546]:
```suggestion
        Gets all voice states of the guild.

```

File: interactions/api/models/guild.py [hunk 22]
```diff
@@ -540,6 +540,38 @@ async def ban(
...
+        states: List[VoiceState] = []
+
+        data = self._client.cache[VoiceState].values.values()
+        states.extend(state for state in data if state.guild_id == self.id)
+        return states
+
+    @property
+    def mapped_voice_states(self) -> Dict[int, List["VoiceState"]]:
+        """
+        Returns all the voice states mapped after their channel id.
```
Comment (Damego) [inline interactions/api/models/guild.py:565]:
```suggestion
        Returns all the voice states mapped after their channel id.

```

**Review by Damego — COMMENTED** [review]
oop, i forgot these

**Review by Damego — APPROVED** [review]
Looks good to me!

**Pull request merged**